compiler/rustc_middle/src/ty/util.rs

```diff
@@ -808,6 +808,26 @@ impl<'tcx> TyCtxt<'tcx> {
             _ => def_kind.article(),
         }
     }
+
+    /// Return `true` if the supplied `CrateNum` is "user-visible," meaning either a [public]
+    /// dependency, or a [direct] private dependency. This is used to decide whether the crate can
+    /// be shown in `impl` suggestions.
+    ///
+    /// [public]: TyCtxt::is_private_dep
+    /// [direct]: rustc_session::cstore::ExternCrate::is_direct
+    pub fn is_user_visible_dep(self, key: CrateNum) -> bool {
+        // | Private | Direct | Visible |                    |
+        // |---------|--------|---------|--------------------|
+        // | Yes     | Yes    | Yes     | !(true && !true)   |
+        // | No      | Yes    | Yes     | !(false && !true)  |
+        // | Yes     | No     | No      | !(true && !false)  |
+        // | No      | No     | Yes     | !(false && !false) |
+        !(self.is_private_dep(key)
+            // If `extern_crate` is `None`, then the crate was injected (e.g., by the allocator).
+            // Treat that kind of crate as "indirect", since it's an implementation detail of
+            // the language.
+            && !self.extern_crate(key.as_def_id()).map_or(false, |e| e.is_direct()))
```

Comment:
```suggestion
        // | Yes     | Yes    | Yes     | !true || true   |
        // | No      | Yes    | Yes     | !false || true  |
        // | Yes     | No     | No      | !true || false  |
        // | No      | No     | Yes     | !false || false |
        !self.is_private_dep(key)
            // If `extern_crate` is `None`, then the crate was injected (e.g., by the allocator).
            // Treat that kind of crate as "indirect", since it's an implementation detail of
            // the language.
            || self.extern_crate(key.as_def_id()).map_or(false, |e| e.is_direct()))
```